### ISSUE TITLE
the definitions of unsupervised and semi-supervised approaches

### ISSUE THREAD

**Issue by Pengyu-Li-2002** [TheDatumOrg/TSB-AD#6]
Dear Author,

Thank you for your open-source work; it has significantly advanced the field of time series anomaly detection (TSAD). However, I have some questions. In the code, you use the training set for both training and validation, and the test set solely for inference, considering this approach as a semi-supervised method. Does this differ from the current definitions of unsupervised and semi-supervised approaches commonly used in the TSAD field? I look forward to your response.

Thank you!

**Comment by qhliu26** [comment]
Hello Pengyu,

Thank you for your interest in our work! I think this is a common practice in this field. Actually, the _test set_ is the entire time series, and the _training set_ is the initial portion of the time series which exhibits no anomaly or very few anomalies.

Here I attach some definitions of unsupervised/semi-supervised approaches in TSAD for reference:
(i) unsupervised, which does not require any labeled data. So we directly pass the _test set_  into the anomaly detectors.
(ii) semi-supervised, requiring labels only for normal instances. So, we use the _training set_ to fix/train the model before applying it to the _test set_.

Let me know if I address your questions and feel free to offer suggestions!

**Comment by nuhdv** [comment]
Dear Author,
I want to know why using the entire time series as a test set, rather than the rest of the time series with the training data removed, differs from the general TSAD setup (i.e., separate training sets for training and separate test sets for testing).
Thank you!

**Comment by FTWBTS** [comment]
Dear Author,

I also believe that your work is highly valuable and has contributed significantly to advancing this field. However, I have a question: In the line train_index = args.filename.split('.')[0].split('_')[-3], some datasets have only 30,000 training samples, while the test set contains 300,000 samples. Wouldn't this create an unfair situation for the model being trained (potential underfitting)?

**Comment by qhliu26** [comment]
Hi! Thanks for bringing up the discussion. In our setup, we adopt a unified evaluation framework where the entire time series is used as test data, and the initial portion of the time series—assumed to be normal or containing a minimal number of anomalies—serves as the training data for semi-supervised methodologies.

The fundamental assumption of semi-supervised approaches, which require training data, is to learn normal patterns from the training dataset and subsequently identify abnormal instances across the entire time series. To achieve this, we assume the initial segment of the time series is either anomaly-free or contains a minimal number of anomalies. However, not all datasets in our collection come pre-split into train/test sets, which provides us the opportunity to recommend a standardized setup.

Specifically, we use the smaller value between the index of the first anomaly and the initial 25% of the time series as the training set across our benchmarking practice. Nevertheless, some datasets present edge cases where the first anomaly appears very early, potentially leading to underfitting issues for learning-based approaches. To mitigate this, we impose a minimum training data length of 500 time steps. In summary, the entire time series is consistently employed as test data, while the initial segment serves as the training data under our unified framework.

Let us know any questions you might have and we are happy to discuss more.

**Comment by AyumiTPK** [comment]
Hi Author,

Thank you for your excellent work on the TSB-AD benchmark and for sharing this powerful framework!

While exploring how semi-supervised and unsupervised models are fairly compared in your paper, I came across your explanation regarding unified evaluation. It really clarified my confusion around how semi-supervised models are applied to the full series, just like unsupervised ones, despite differences in training. 

To ensure I am using your pipeline correctly for my research project (I’m a student, still learning!), would it be possible for you to share the semi-supervised version of the following code snippet?

`import pandas as pd
from TSB_AD.model_wrapper import run_Unsupervise_AD
from TSB_AD.evaluation.metrics import get_metrics

# Specify Anomaly Detector to use and data directory
AD_Name = 'IForest'   # It can be replaced with any anomaly detector availale in TSB-AD
data_direc = 'Datasets/TSB-AD-U/001_NAB_id_1_Facility_tr_1007_1st_2014.csv'

# Loading Data
df = pd.read_csv(data_direc).dropna()
data = df.iloc[:, 0:-1].values.astype(float)
label = df['Label'].astype(int).to_numpy()

# Applying Anomaly Detector
output = run_Unsupervise_AD(AD_Name, data)

# Evaluation
evaluation_result = get_metrics(output, label)`

**Comment by JTstudyhappy** [comment]
你好！想问一下您的意思是，无监督不需要训练，而半监督仍然需要无标签数据来训练的意思吗？
